pacstall/cmds/download.py

```diff
@@ -24,59 +24,159 @@
+    client
+        The asynchronous httpx client.
+    url
+        The URL to download.
+    download_task
+        The task ID of the download task. Used to update the progress bar.
+    progress_bar
+        The progress bar.
+
+    Returns
+    -------
+    int
+        Exit codes (return codes) depending on the status of the download.
+        - ``0`` for success.
+        - :atrr:`pacstall.ErrorCodes.UNAVAILABLE_ERROR` (69) for when there's a
+          :exc:`HTTPStatusError` or :exc:`RequestError`.
+        - :atrr:`ErrorCodes.IO_ERROR` for when there's an :exc:`OSError`.
     """
+    filename = url.split("/")[-1].split(".pacscript")[0]
+
+    # TODO: `log.debug` each successfully downloaded file.
+    #       If verbose level is high enough.
+    try:
+        response = await client.get(url)
+        response.raise_for_status()
+        async with aiofiles.open(Path.cwd() / url.split("/")[-1], mode="wb") as file:
+            await file.write(response.content)
+
+        progress_bar.update(download_task, advance=1, filename=filename)
+
+        return 0
+
+    except HTTPStatusError as error:
+        rprint(
+            f"[bold red]Failed[/bold red]: {filename} (HTTP status error: {error.response.status_code})"
+        )
+        log = getLogger()
+        log.debug(f"Failed: {filename}", exc_info=True)
+        return ErrorCodes.UNAVAILABLE_ERROR
+
+    except RequestError as error:
+        error_message = str(error) or type(error).__name__
+        rprint(f"[bold red]Failed[/bold red]: {filename} ({error_message})")
+
+        log = getLogger()
+        log.debug(f"Failed: {filename}", exc_info=True)
+        return ErrorCodes.UNAVAILABLE_ERROR
+
+    except OSError as error:
+        error_message = str(error) or type(error).__name__
+        rprint(f"[bold red]Failed[/bold red]: {filename} (OS error: {error_message}")
+
+        log = getLogger()
+        log.debug(f"Failed: {filename}", exc_info=True)
+        return ErrorCodes.IO_ERROR
 
-    REQUEST_ERROR_MESSAGES = {
-        exceptions.HTTPError: "A HTTP error occurred while connecting to the URL",
-        exceptions.ConnectionError: "No internet connection detected",
-        exceptions.Timeout: "Connection timed out. Check your internet connection",
-        exceptions.TooManyRedirects: "Too many redirections. Possibly bad URL",
+
+async def execute(pacscripts: List[str]) -> int:
+    """
+    Executes the download command.
+
+    Parameters
+    ----------
+    pacscripts
+        The pacscripts to download.
+
+    Returns
+    -------
+    int
+        Returns the greatest returned exit code from :func:`download`
+
+    See Also
+    --------
+    download : Asynchronous download function for the pacscript URLs.
+
+    Notes
+    -----
+    This function currently only supports downloading pacscripts from the
+    `official repository <https://github.com/pacstall/pacstall-programs>`_.
+    """
+    # TODO: Add ability to download from other repositories.
```

Comment:
My suggestion would be using a class or a tuple that stores the package name, repo name/url, and the information if it's being explicitly installed or is a dependency, and use the same format everywhere we need it